scripts/undo.pl

```diff
@@ -0,0 +1,186 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+use 5.10.1;
+use strict;
+use warnings;
+
+use File::Basename qw(dirname);
+use File::Spec::Functions qw(catfile catdir rel2abs);
+use Cwd qw(realpath);
+BEGIN {
+    require lib;
+    my $dir = rel2abs(catdir(dirname(__FILE__), '..'));
+    lib->import($dir, catdir($dir, 'lib'), catdir($dir, qw(local lib perl5)));
+}
+
+use Bugzilla;
+BEGIN { Bugzilla->extensions };
+
+use Try::Tiny;
+
+# TODO: we'll have many calls to undo() here, per instructions from Erin & Dave Camp.
+my $bug_id = 344674;
+
+
+# not involved with kmag
+my $query = q{
+    resolution = 'INACTIVE'
+    AND NOT ( NOT ( triage_owner.userid = ? )
+    AND NOT ( assigned_to = ?)
+    AND NOT ( reporter = ?)
+    AND NOT (qa_contact IS NOT NULL AND qa_contact = ?)
+    AND NOT (
+        triage_owner.login_name = ?
+        AND bugs.priority IN ('P4', 'P5')
+    ))
+};
+undo($query, (106098) x 4, 'mak77@bonardo.net');
+
+sub undo {
+    my $changes = get_changes(@_);
+    my $comments = get_comments(@_);
+
+    my %action;
+    while ($_ = $changes->()) {
+        push @{ $action{$_->{bug_id}}{$_->{bug_when}}{remove_activities} }, { id => $_->{change_id} };
+        $action{ $_->{bug_id} }{ $_->{bug_when} }{change}{ $_->{field_name} } = {
+            replace => $_->{added},
+            with    => $_->{removed},
+        };
+    }
+
+    while ($_ = $comments->()) {
+        push @{ $action{ $_->{bug_id} }{$_->{bug_when}}{remove_comments} }, {
+            id => $_->{comment_id},
+        };
+    }
+
+    my $dbh = Bugzilla->dbh;
+    foreach my $bug_id (keys %action) {
+        $dbh->bz_start_transaction;
+        say "working on $bug_id";
+        try {
+            my ($delta_ts, $lastdiffed) = $dbh->selectrow_array(
+                'SELECT delta_ts, lastdiffed FROM bugs WHERE bug_id = ?',
+                undef,
+                $bug_id);
+            my ($previous_last_ts) = $dbh->selectrow_array(
+                'SELECT bug_when FROM bugs_activity WHERE bug_when < ? ORDER BY bug_when DESC LIMIT 1',
+                undef,
+                $delta_ts
+            );
+            die "cannot find previous last updated time" unless $previous_last_ts;
+            my $action = delete $action{$bug_id}{$delta_ts};
+            if (keys %{ $action{$bug_id}{$delta_ts}}) {
+                die "skipping because more than one change\n";
+            }
+            elsif (!$action) {
+                die "skipping because most recent change newer than automation change\n";
+            }
+            $action->{change}{delta_ts} = { replace => $delta_ts, with => $previous_last_ts };
+            $action->{change}{lastdiffed} = { replace => $delta_ts, with => $previous_last_ts };
```

Comment:
As you also mentioned in IRC, these need to actually be updated in the DB for the bugs table.

Reply:
oh wait nm. i see that they ARE updated.